server/src/main/java/io/envoyproxy/controlplane/server/serializer/DefaultProtoResourcesSerializer.java

```diff
@@ -12,7 +14,34 @@
    * {@inheritDoc}
    */
   @Override
-  public Any serialize(Message resource) {
-    return Any.pack(resource);
+  public Any serialize(Message resource, ApiVersion apiVersion) {
+    Any output = Any.pack(resource);
+
+    return maybeRewriteTypeUrl(output, apiVersion);
+  }
+
+  protected Any maybeRewriteTypeUrl(Any output, ApiVersion apiVersion) {
+    // If the requested version and output version match, we can just return the output.
+    if (Resources.getResourceApiVersion(output.getTypeUrl()) == apiVersion) {
+      return output;
+    }
+
+    // Rewrite the type URL to the requested version. This takes advantage of the fact that v2
+    // and v3 resources are wire compatible (as long as callers are careful about deprecated v2
+    // fields or new v3 fields).
+    // TODO: revisit this if https://github.com/envoyproxy/envoy/issues/10776 is implemented which
+    //  would mean we don't need to do any type URL rewriting since Envoy can handle either.
+    switch (apiVersion) {
+      case V2:
+        return output.toBuilder()
+            .setTypeUrl(Resources.V3_TYPE_URLS_TO_V2.get(output.getTypeUrl()))
+            .build();
+      case V3:
+        return output.toBuilder()
+            .setTypeUrl(Resources.V2_TYPE_URLS_TO_V3.get(output.getTypeUrl()))
+            .build();
```

Comment:
I added conversion in this PR. the type URL of the output Any will be rewritten to match the request version if they're different.

This is covered in V2DiscoveryServerV3ResourcesXdsIT, V2DiscoveryServerV3ResourcesAdsIT, V3DiscoveryServerV2ResourcesXdsIT, and V3DiscoveryServerV2ResourcesAdsIT.

Those tests do not cover using resource_api_version at all, which I'd be open to adding if folks think there'd be value